rust/otap-dataflow/crates/otap/src/syslog_cef_receiver.rs

```diff
@@ -110,43 +123,63 @@ pub static SYSLOG_CEF_RECEIVER: ReceiverFactory<OtapPdata> = ReceiverFactory {
+        let mut telemetry_timer = Some(handle);
+
         match self.config.protocol {
             Protocol::Tcp => {
                 let listener = effect_handler.tcp_listener(self.config.listening_addr)?;
 
                 loop {
                     tokio::select! {
-                        biased; //Prioritize control messages over data
+                        biased; // Prioritize control messages over data
 
                         // Process incoming control messages.
                         ctrl_msg = ctrl_chan.recv() => {
                             match ctrl_msg {
-                                Ok(NodeControlMsg::Shutdown {..}) => {
-                                // ToDo: Add proper deadline function
-                                break;
-                                },
-                            Err(e) => {
-                                return Err(Error::ChannelRecvError(e));
+                                Ok(NodeControlMsg::Shutdown { .. }) => {
+                                    if let Some(h) = telemetry_timer.take() {
+                                        let _ = h.cancel().await;
+                                    }
+                                    break;
+                                }
+                                Ok(NodeControlMsg::CollectTelemetry { mut metrics_reporter }) => {
+                                    let mut guard = self.metrics.borrow_mut();
+                                    let _ = metrics_reporter.report(&mut guard);
                                 }
-                            _ => {
-                                // ToDo: Handle other control messages if needed
+                                Err(e) => {
+                                    return Err(Error::ChannelRecvError(e));
+                                }
+                                _ => {
+                                    // ToDo: Handle other control messages if needed
                                 }
                             }
                         }
 
+
                         // Process incoming TCP connections.
                         accept_result = listener.accept() => {
                             match accept_result {
                                 Ok((socket, _peer_addr)) => {
-                                    // Clone the effect handler so the spawned task can send messages.
+                                    // Track active connections
+                                    {
+                                        let mut m = self.metrics.borrow_mut();
+                                        m.tcp_connections_active.inc();
+                                    }
+
+                                    // Clone the effect handler and metrics so the spawned task can send messages and record telemetry.
                                     let effect_handler = effect_handler.clone();
+                                    let metrics = self.metrics.clone();
```

Comment:
We don't want to clone the `MetricSet`.

Reply:
i have updated this to be obvious we are cloning the Rc pointer not the MetricSet itself

Reply:
Ah that's my bad! I mistakenly assumed that it's cloning the `MetricSet`.